tests/neo-vm.Tests/UtExecutionEngine.cs

```diff
@@ -0,0 +1,104 @@
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Neo.VM;
+
+namespace Neo.Test
+{
+    [TestClass]
+    public class UtExecutionEngine
+    {
+        [TestMethod]
+        public void TestReferenceTracing()
+        {
+            using ScriptBuilder sb = new ScriptBuilder();
+            sb.EmitPush(0); //{0}{}:1
+            sb.Emit(OpCode.NEWARRAY); //{A[]}|{}:1
+            sb.Emit(OpCode.DUP); //{A[],A[]}|{}:2
+            sb.Emit(OpCode.DUP); //{A[],A[],A[]}|{}:3
+            sb.Emit(OpCode.APPEND); //{A[A]}|{}:2
+            sb.Emit(OpCode.DUP); //{A[A],A[A]}|{}:3
+            sb.EmitPush(0); //{A[A],A[A],0}|{}:4
+            sb.Emit(OpCode.NEWARRAY); //{A[A],A[A],B[]}|{}:4
+            sb.Emit(OpCode.TOALTSTACK); //{A[A],A[A]}|{B[]}:4
+            sb.Emit(OpCode.DUPFROMALTSTACK); //{A[A],A[A],B[]}|{B[]}:5
+            sb.Emit(OpCode.APPEND); //{A[A,B]}|{B[]}:4
+            sb.Emit(OpCode.DUPFROMALTSTACK); //{A[A,B],B[]}|{B[]}:5
+            sb.EmitPush(0); //{A[A,B],B[],0}|{B[]}:6
+            sb.Emit(OpCode.NEWARRAY); //{A[A,B],B[],C[]}|{B[]}:6
+            sb.Emit(OpCode.TUCK); //{A[A,B],C[],B[],C[]}|{B[]}:7
+            sb.Emit(OpCode.APPEND); //{A[A,B],C[]}|{B[C]}:6
+            sb.EmitPush(0); //{A[A,B],C[],0}|{B[C]}:7
+            sb.Emit(OpCode.NEWARRAY); //{A[A,B],C[],D[]}|{B[C]}:7
+            sb.Emit(OpCode.TUCK); //{A[A,B],D[],C[],D[]}|{B[C]}:8
+            sb.Emit(OpCode.APPEND); //{A[A,B],D[]}|{B[C[D]]}:7
+            sb.Emit(OpCode.DUPFROMALTSTACK); //{A[A,B],D[],B[C]}|{B[C[D]]}:8
+            sb.Emit(OpCode.APPEND); //{A[A,B]}|{B[C[D[B]]]}:7
+            sb.Emit(OpCode.FROMALTSTACK); //{A[A,B],B[C[D[B]]]}:7
+            sb.Emit(OpCode.DROP); //{A[A,B[C[D[B]]]]}:6
+            sb.Emit(OpCode.DUP); //{A[A,B[C[D[B]]]],A[A,B]}:7
+            sb.EmitPush(1); //{A[A,B[C[D[B]]]],A[A,B],1}:8
+            sb.Emit(OpCode.REMOVE); //{A[A]}:2
+            sb.Emit(OpCode.DROP); //{}:0
+            sb.Emit(OpCode.RET);
+            using ExecutionEngine engine = new ExecutionEngine();
+            Debugger debugger = new Debugger(engine);
+            engine.LoadScript(sb.ToArray());
+            Assert.AreEqual(VMState.BREAK, debugger.StepInto());
+            Assert.AreEqual(1, engine.StackItemCount);
+            Assert.AreEqual(VMState.BREAK, debugger.StepInto());
+            Assert.AreEqual(1, engine.StackItemCount);
+            Assert.AreEqual(VMState.BREAK, debugger.StepInto());
+            Assert.AreEqual(2, engine.StackItemCount);
```

Comment:
Why `DUP` Increase the item count?

Reply:
Of course. DUP add an item to the stack.

Reply:
Before the change, it only sum the different items `.Distinct()`

https://github.com/neo-project/neo-vm/blob/335a9eff86367ad206b7309ab0b7094448e7d2c6/src/neo-vm/ExecutionEngine.cs#L114

Reply:
I am doing the UT with master, and it pass, so i have something wrong xD i will review it

Reply:
The `Distinct` is applied on the stacks, not on the items.

Reply:
Could you explain me the rules of `GetItemCount`?

new Array(1) => Count as 2 (GOOD)
Now we `DUP` and the count is `3`?
But we have 2 Arrays( dupped ) both with 1 entry.

In memory we only have 2 items, the array and the entry, the others are references to this array. So I can understand 2, or 4. But not 3 :P

Reply:
2 reference to the array, and 1 to the element.

Reply:
but if we mix references with elements, it's weird, reference will be 8 bytes in x64 (the pointer), and the element could be bigger

Reply:
If references are not counted, you can create unlimited references and fill up the memory.

Reply:
That's the reason why I think that we should take into account the memory. Now you can have two arrays of 1024, but not one of 1025.

Reply:
We can limit the `MaxStackSize` only, and remove `MaxArraySize`.